// @airgap/beacon-sdk/dist/cjs/transports/P2PTransport.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.P2PTransport = void 0;
var Logger_1 = require("../utils/Logger");
var __1 = require("..");
var events_1 = require("../events");
var PeerManager_1 = require("../managers/PeerManager");
var logger = new Logger_1.Logger('Transport');
var P2PTransport = /** @class */ (function (_super) {
    __extends(P2PTransport, _super);
    function P2PTransport(name, keyPair, storage, events, isDapp) {
        var _this = _super.call(this, name) || this;
        _this.type = __1.TransportType.P2P;
        _this.isDapp = true;
        // Make sure we only listen once
        _this.listeningForChannelOpenings = false;
        _this.keyPair = keyPair;
        _this.storage = storage;
        _this.events = events;
        _this.isDapp = isDapp;
        _this.client = new __1.P2PCommunicationClient(_this.name, _this.keyPair, 1, false);
        _this.peerManager = new PeerManager_1.PeerManager(storage);
        return _this;
    }
    P2PTransport.isAvailable = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, Promise.resolve(true)];
            });
        });
    };
    P2PTransport.prototype.connect = function () {
        return __awaiter(this, void 0, void 0, function () {
            var knownPeers, connectionPromises;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        logger.log('connect');
                        this._isConnected = __1.TransportStatus.CONNECTING;
                        return [4 /*yield*/, this.client.start()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.peerManager.getPeers()];
                    case 2:
                        knownPeers = _a.sent();
                        if (!(knownPeers.length > 0)) return [3 /*break*/, 4];
                        logger.log('connect', "connecting to " + knownPeers.length + " peers");
                        connectionPromises = knownPeers.map(function (peer) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                            return [2 /*return*/, this.listen(peer.publicKey)];
                        }); }); });
                        return [4 /*yield*/, Promise.all(connectionPromises)];
                    case 3:
                        _a.sent();
                        return [3 /*break*/, 6];
                    case 4:
                        if (!this.isDapp) return [3 /*break*/, 6];
                        return [4 /*yield*/, this.connectNewPeer()];
                    case 5:
                        _a.sent();
                        _a.label = 6;
                    case 6: return [4 /*yield*/, _super.prototype.connect.call(this)];
                    case 7:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PTransport.prototype.reconnect = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.isDapp) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.connectNewPeer()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    P2PTransport.prototype.connectNewPeer = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                logger.log('connectNewPeer');
                return [2 /*return*/, new Promise(function (resolve) { return __awaiter(_this, void 0, void 0, function () {
                        var _a, _b, _c;
                        var _this = this;
                        return __generator(this, function (_d) {
                            switch (_d.label) {
                                case 0:
                                    if (!!this.listeningForChannelOpenings) return [3 /*break*/, 2];
                                    return [4 /*yield*/, this.client.listenForChannelOpening(function (publicKey) { return __awaiter(_this, void 0, void 0, function () {
                                            var newPeer;
                                            return __generator(this, function (_a) {
                                                switch (_a.label) {
                                                    case 0:
                                                        logger.log('connectNewPeer', "new publicKey " + publicKey);
                                                        newPeer = { name: '', publicKey: publicKey, relayServer: '' };
                                                        return [4 /*yield*/, this.peerManager.hasPeer(publicKey)];
                                                    case 1:
                                                        if (!!(_a.sent())) return [3 /*break*/, 4];
                                                        return [4 /*yield*/, this.peerManager.addPeer(newPeer)];
                                                    case 2:
                                                        _a.sent();
                                                        return [4 /*yield*/, this.listen(publicKey)];
                                                    case 3:
                                                        _a.sent();
                                                        _a.label = 4;
                                                    case 4:
                                                        this.events
                                                            .emit(events_1.BeaconEvent.P2P_CHANNEL_CONNECT_SUCCESS, newPeer)
                                                            .catch(function (emitError) { return console.warn(emitError); });
                                                        resolve();
                                                        return [2 /*return*/];
                                                }
                                            });
                                        }); })];
                                case 1:
                                    _d.sent();
                                    this.listeningForChannelOpenings = true;
                                    _d.label = 2;
                                case 2:
                                    _b = (_a = this.events).emit;
                                    _c = [events_1.BeaconEvent.P2P_LISTEN_FOR_CHANNEL_OPEN];
                                    return [4 /*yield*/, this.client.getHandshakeInfo()];
                                case 3:
                                    _b.apply(_a, _c.concat([_d.sent()]))
                                        .catch(function (emitError) { return console.warn(emitError); });
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            });
        });
    };
    P2PTransport.prototype.getPeers = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.peerManager.getPeers()];
            });
        });
    };
    P2PTransport.prototype.addPeer = function (newPeer) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.peerManager.hasPeer(newPeer.publicKey)];
                    case 1:
                        if (!!(_a.sent())) return [3 /*break*/, 5];
                        logger.log('addPeer', newPeer);
                        return [4 /*yield*/, this.peerManager.addPeer({
                                name: newPeer.name,
                                publicKey: newPeer.publicKey,
                                relayServer: newPeer.relayServer
                            })];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, this.client.openChannel(newPeer.publicKey, newPeer.relayServer)]; // TODO: Should we have a confirmation here?
                    case 3:
                        _a.sent(); // TODO: Should we have a confirmation here?
                        return [4 /*yield*/, this.listen(newPeer.publicKey)]; // TODO: Prevent channels from being opened multiple times
                    case 4:
                        _a.sent(); // TODO: Prevent channels from being opened multiple times
                        return [3 /*break*/, 6];
                    case 5:
                        logger.log('addPeer', 'peer already added, skipping', newPeer);
                        _a.label = 6;
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    P2PTransport.prototype.removePeer = function (peerToBeRemoved) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        logger.log('removePeer', peerToBeRemoved);
                        return [4 /*yield*/, this.peerManager.removePeer(peerToBeRemoved.publicKey)];
                    case 1:
                        _a.sent();
                        if (!this.client) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.client.unsubscribeFromEncryptedMessage(peerToBeRemoved.publicKey)];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    P2PTransport.prototype.removeAllPeers = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        logger.log('removeAllPeers');
                        return [4 /*yield*/, this.peerManager.removeAllPeers()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.client.unsubscribeFromEncryptedMessages()];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PTransport.prototype.send = function (message, recipient) {
        return __awaiter(this, void 0, void 0, function () {
            var knownPeers, promises;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.storage.get(__1.StorageKey.TRANSPORT_P2P_PEERS)];
                    case 1:
                        knownPeers = _a.sent();
                        if (!recipient) return [3 /*break*/, 2];
                        if (!knownPeers.some(function (peer) { return peer.publicKey === recipient; })) {
                            throw new Error('Recipient unknown');
                        }
                        return [2 /*return*/, this.client.sendMessage(recipient, message)];
                    case 2:
                        promises = knownPeers.map(function (peer) { return _this.client.sendMessage(peer.publicKey, message); });
                        return [4 /*yield*/, Promise.all(promises)];
                    case 3: return [2 /*return*/, (_a.sent())[0]];
                }
            });
        });
    };
    P2PTransport.prototype.listen = function (publicKey) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client
                            .listenForEncryptedMessage(publicKey, function (message) {
                            var connectionContext = {
                                origin: __1.Origin.P2P,
                                id: publicKey
                            };
                            _this.notifyListeners(message, connectionContext).catch(function (error) {
                                throw error;
                            });
                        })
                            .catch(function (error) {
                            throw error;
                        })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return P2PTransport;
}(__1.Transport));
exports.P2PTransport = P2PTransport;
