// @airgap/beacon-sdk/dist/cjs/transports/Transport.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transport = void 0;
var Logger_1 = require("../utils/Logger");
var __1 = require("..");
var logger = new Logger_1.Logger('Transport');
var Transport = /** @class */ (function () {
    function Transport(name) {
        this.type = __1.TransportType.POST_MESSAGE;
        this._isConnected = __1.TransportStatus.NOT_CONNECTED;
        this.listeners = [];
        this.name = name;
    }
    Object.defineProperty(Transport.prototype, "connectionStatus", {
        get: function () {
            return this._isConnected;
        },
        enumerable: false,
        configurable: true
    });
    Transport.isAvailable = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, Promise.resolve(false)];
            });
        });
    };
    Transport.prototype.connect = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                logger.log('connect');
                this._isConnected = __1.TransportStatus.CONNECTED;
                return [2 /*return*/];
            });
        });
    };
    Transport.prototype.reconnect = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                logger.log('reconnect');
                return [2 /*return*/];
            });
        });
    };
    Transport.prototype.send = function (message, recipient) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                logger.log('send', message, recipient);
                return [2 /*return*/];
            });
        });
    };
    Transport.prototype.addListener = function (listener) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                logger.log('addListener');
                this.listeners.push(listener);
                return [2 /*return*/];
            });
        });
    };
    Transport.prototype.removeListener = function (listener) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                logger.log('removeListener');
                this.listeners = this.listeners.filter(function (element) { return element !== listener; });
                return [2 /*return*/];
            });
        });
    };
    Transport.prototype.notifyListeners = function (message, connectionInfo) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                logger.log('notifyListeners');
                if (this.listeners.length === 0) {
                    logger.warn('notifyListeners', '0 listeners notified!', this);
                }
                else {
                    logger.log("Notifying " + this.listeners.length + " listeners", this);
                }
                this.listeners.forEach(function (listener) {
                    listener(message, connectionInfo);
                });
                return [2 /*return*/];
            });
        });
    };
    return Transport;
}());
exports.Transport = Transport;
