// @airgap/beacon-sdk/dist/cjs/clients/wallet-client/WalletClient.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletClient = void 0;
var __1 = require("../..");
var PermissionManager_1 = require("../../managers/PermissionManager");
var AppMetadataManager_1 = require("../../managers/AppMetadataManager");
var IncomingBeaconMessageInterceptor_1 = require("../../interceptors/IncomingBeaconMessageInterceptor");
var OutgoingBeaconMessageInterceptor_1 = require("../../interceptors/OutgoingBeaconMessageInterceptor");
var WalletClient = /** @class */ (function (_super) {
    __extends(WalletClient, _super);
    function WalletClient(config) {
        var _this = _super.call(this, { name: config.name, storage: new __1.LocalStorage() }) || this;
        _this.pendingRequests = [];
        _this.permissionManager = new PermissionManager_1.PermissionManager(new __1.LocalStorage());
        _this.appMetadataManager = new AppMetadataManager_1.AppMetadataManager(new __1.LocalStorage());
        return _this;
    }
    WalletClient.prototype.init = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, _super.prototype.init.call(this, false)];
            });
        });
    };
    WalletClient.prototype.connect = function (newMessageCallback) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.handleResponse = function (message, connectionInfo) { return __awaiter(_this, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                if (!!this.pendingRequests.some(function (request) { return request.id === message.id; })) return [3 /*break*/, 2];
                                this.pendingRequests.push(message);
                                return [4 /*yield*/, IncomingBeaconMessageInterceptor_1.IncomingBeaconMessageInterceptor.intercept({
                                        message: message,
                                        connectionInfo: connectionInfo,
                                        appMetadataManager: this.appMetadataManager,
                                        interceptorCallback: newMessageCallback
                                    })];
                            case 1:
                                _a.sent();
                                _a.label = 2;
                            case 2: return [2 /*return*/];
                        }
                    });
                }); };
                return [2 /*return*/, _super.prototype._connect.call(this)];
            });
        });
    };
    WalletClient.prototype.respond = function (message) {
        return __awaiter(this, void 0, void 0, function () {
            var request, _a, _b, _c;
            var _this = this;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        console.log('responding to message', message);
                        request = this.pendingRequests.find(function (pendingRequest) { return pendingRequest.id === message.id; });
                        if (!request) {
                            throw new Error('No matching request found!');
                        }
                        this.pendingRequests = this.pendingRequests.filter(function (pendingRequest) { return pendingRequest.id !== message.id; });
                        _b = (_a = OutgoingBeaconMessageInterceptor_1.OutgoingBeaconMessageInterceptor).intercept;
                        _c = {};
                        return [4 /*yield*/, this.beaconId];
                    case 1: return [4 /*yield*/, _b.apply(_a, [(_c.beaconId = _d.sent(),
                                _c.request = request,
                                _c.message = message,
                                _c.permissionManager = this.permissionManager,
                                _c.appMetadataManager = this.appMetadataManager,
                                _c.interceptorCallback = function (response) { return __awaiter(_this, void 0, void 0, function () {
                                    return __generator(this, function (_a) {
                                        switch (_a.label) {
                                            case 0: return [4 /*yield*/, this.respondToMessage(response)];
                                            case 1:
                                                _a.sent();
                                                return [2 /*return*/];
                                        }
                                    });
                                }); },
                                _c)])];
                    case 2:
                        _d.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    WalletClient.prototype.getAppMetadataList = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.appMetadataManager.getAppMetadataList()];
            });
        });
    };
    WalletClient.prototype.getAppMetadata = function (beaconId) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.appMetadataManager.getAppMetadata(beaconId)];
            });
        });
    };
    WalletClient.prototype.removeAppMetadata = function (beaconId) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.appMetadataManager.removeAppMetadata(beaconId)];
            });
        });
    };
    WalletClient.prototype.removeAllAppMetadata = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.appMetadataManager.removeAllAppMetadata()];
            });
        });
    };
    WalletClient.prototype.getPermissions = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.permissionManager.getPermissions()];
            });
        });
    };
    WalletClient.prototype.getPermission = function (accountIdentifier) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.permissionManager.getPermission(accountIdentifier)];
            });
        });
    };
    WalletClient.prototype.removePermission = function (accountIdentifier) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.permissionManager.removePermission(accountIdentifier)];
            });
        });
    };
    WalletClient.prototype.removeAllPermissions = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.permissionManager.removeAllPermissions()];
            });
        });
    };
    WalletClient.prototype.removePeer = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            var removePeerResult;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.transport];
                    case 1:
                        if (!((_a.sent()).type === __1.TransportType.P2P)) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.transport];
                    case 2:
                        removePeerResult = (_a.sent()).removePeer(id);
                        return [4 /*yield*/, this.removePermissionsForPeers([id])];
                    case 3:
                        _a.sent();
                        return [2 /*return*/, removePeerResult];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    WalletClient.prototype.removeAllPeers = function () {
        return __awaiter(this, void 0, void 0, function () {
            var peers, removePeerResult;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.transport];
                    case 1:
                        if (!((_a.sent()).type === __1.TransportType.P2P)) return [3 /*break*/, 6];
                        return [4 /*yield*/, this.transport];
                    case 2: return [4 /*yield*/, (_a.sent()).getPeers()];
                    case 3:
                        peers = _a.sent();
                        return [4 /*yield*/, this.transport];
                    case 4:
                        removePeerResult = (_a.sent()).removeAllPeers();
                        return [4 /*yield*/, this.removePermissionsForPeers(peers)];
                    case 5:
                        _a.sent();
                        return [2 /*return*/, removePeerResult];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    WalletClient.prototype.removePermissionsForPeers = function (peersToRemove) {
        return __awaiter(this, void 0, void 0, function () {
            var permissions, peerIdsToRemove, permissionsToRemove, permissionIdentifiersToRemove;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.permissionManager.getPermissions()];
                    case 1:
                        permissions = _a.sent();
                        peerIdsToRemove = peersToRemove.map(function (peer) { return peer.publicKey; });
                        permissionsToRemove = permissions.filter(function (permission) {
                            return peerIdsToRemove.includes(permission.appMetadata.beaconId);
                        });
                        permissionIdentifiersToRemove = permissionsToRemove.map(function (permissionInfo) { return permissionInfo.accountIdentifier; });
                        return [4 /*yield*/, this.permissionManager.removePermissions(permissionIdentifiersToRemove)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    WalletClient.prototype.respondToMessage = function (message) {
        return __awaiter(this, void 0, void 0, function () {
            var serializedMessage;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, new __1.Serializer().serialize(message)];
                    case 1:
                        serializedMessage = _a.sent();
                        return [4 /*yield*/, this.transport];
                    case 2: return [4 /*yield*/, (_a.sent()).send(serializedMessage)];
                    case 3:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return WalletClient;
}(__1.Client));
exports.WalletClient = WalletClient;
